test/anki_viewer/rule_test.exs

```diff
@@ -0,0 +1,117 @@
+defmodule AnkiViewer.RuleTest do
+  use AnkiViewer.DataCase, async: false
+
+  test "insert rule struct" do
+    attrs = %{
+      name: "no blank fields",
+      code: """
+      (_deck, note) => note.sfld !== ""
+      """,
+      tests:
+        [
+          %{
+            note: %{sfld: "h"},
+            fine: true
+          },
+          %{
+            note: %{sfld: ""},
+            fine: false
+          }
+        ]
+        |> Jason.encode!()
+    }
+
+    %Rule{}
+    |> Map.merge(attrs)
+    |> Rule.insert!()
+  end
+
+  describe "rule" do
+    setup do
+      attrs = [
+        %{
+          name: "no duplicate front",
+          code: """
+          (deck, note) => {
+            var fine = true;
+            for (var i=0;i<deck.length;i++) {
+              if (deck[i].sfld == note.sfld && deck[i].nid != note.nid) {
+                fine = false;
+                break;
+              }
+            }
+            return fine;
+          }
+          """,
+          tests: """
+          [
+            {
+              "deck": [
+                {"nid": 0, "sfld": "h"},
+                {"nid": 1, "sfld": "hi"}
+              ],
+              "note": {"nid": 0, "sfld": "h"},
+              "fine": true
+            },
+            {
+              "deck": [
+                {"nid": 0, "sfld": "h"},
+                {"nid": 1, "sfld": "h"}
+              ],
+              "note": {"nid": 0, "sfld": "h"},
+              "fine": true
+            }
+          ]
+          """
```

Comment:
I would prefer this to be elixir syntax which we pass through `Jason.encode!` (keeping consistency and maintainability)